apps/sync/js/sync-engine/syncengine.js

```diff
@@ -0,0 +1,358 @@
+ *     up to the remote Firefox Sync server (again, through the remote Syncto
+ *     proxy server). The RemoteTransformer on the Kinto collection will take
+ *     care of encrypting the records before they get uploaded.
+ *   - Call the `handleConflict` method on the DataAdapter again for any
+ *     conflicts kinto.js may report during this second sync process.
+ *
+ */
+
+'use strict';
+
+/* global
+  crypto,
+  FxSyncWebCrypto,
+  Kinto
+*/
+
+/* exported
+  SyncEngine
+*/
+
+var SyncEngine = (function() {
+  var FxSyncIdSchema = Kinto.createIdSchema({
+    constructor: function(collectionName) {
+      this.collectionName = collectionName;
+    },
+    generate: function() {
+      var bytes = new Uint8Array(9);
+      crypto.getRandomValues(bytes);
+      var binStr = '';
+      for (var i = 0; i < 9; i++) {
+          binStr += String.fromCharCode(bytes[i]);
+      }
+      // See https://docs.services.mozilla.com/storage/apis-1.5.html
+      return window.btoa(binStr).replace('+', '-').replace('/', '_');
+    },
+    validate: function(id) {
+      // FxSync id's should be 12 ASCII characters, representing 9 bytes of data
+      // in modified Base64 for URL variants exist, where the '+' and '/'
+      // characters of standard Base64 are respectively replaced by '-' and '_'
+      // See https://docs.services.mozilla.com/storage/apis-1.5.html
+      return /^[A-Za-z0-9-_]{12}$/.test(id);
+    }
+  });
+
+  var ControlCollectionIdSchema = Kinto.createIdSchema({
+    constructor: function(collectionName, keyName) {
+      this.collectionName = collectionName;
+      this.keyName = keyName;
+    },
+    generate: function() {
+      return this.keyName;
+    },
+    validate: function(id) {
+      return (id === this.keyName);
+    }
+  });
+
+  var WebCryptoTransformer = Kinto.createRemoteTransformer({
+    constructor: function(collectionName, fswc) {
+      if (!fswc.bulkKeyBundle) {
+        throw new Error(`Attempt to register Transformer with no bulk key bundl\
+e!`);
+      }
+      this.collectionName = collectionName;
+      this.fswc = fswc;
+    },
+    encode: function(record) {
+      return this.fswc.encrypt(record.payload, this.collectionName).then(
+          payloadEnc => {
+        record.payload = JSON.stringify(payloadEnc);
+        return record;
+      });
+    },
+    decode: function(record) {
+      // Allowing JSON.parse errors to bubble up to the errors list in the
+      // syncResults:
+      return this.fswc.decrypt(JSON.parse(record.payload), this.collectionName).
+          then(payloadDec => {
+        record.payload = payloadDec;
+        return record;
+      });
+    }
+  });
+
+  /**
+    * SyncEngine - Constructor.
+    * @param {Object} options Should contain the following fields:
+    *                         * URL - e.g. 'http://localhost:8000/v1/'
+    *                         * assertion - a BrowserID assertion (Base64)
+    *                         * xClientState - xClientState header value
```

Comment:
If I am not wrong, we cannot generate xClientState on the System app unless we use some of the crypto tools that the Sync app has. So you'll need to generate it here.